# Laboratorio 4

## Inteligencia Artificial

### Andres de la Roca 20332

In [127]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt

### Task 1.1
Leer el archivo CSV proporcionado y almacenarlo en un np.array para ser trabajado en el notebook.

In [128]:
data = pd.read_csv('framingham.csv')

In [129]:
data

male  age  education  currentSmoker  cigsPerDay  BPMeds  \
0        1   39        4.0              0         0.0     0.0   
1        0   46        2.0              0         0.0     0.0   
2        1   48        1.0              1        20.0     0.0   
3        0   61        3.0              1        30.0     0.0   
4        0   46        3.0              1        23.0     0.0   
...    ...  ...        ...            ...         ...     ...   
4233     1   50        1.0              1         1.0     0.0   
4234     1   51        3.0              1        43.0     0.0   
4235     0   48        2.0              1        20.0     NaN   
4236     0   44        1.0              1        15.0     0.0   
4237     0   52        2.0              0         0.0     0.0   

      prevalentStroke  prevalentHyp  diabetes  totChol  sysBP  diaBP    BMI  \
0                   0             0         0    195.0  106.0   70.0  26.97   
1                   0             0         0    250.0  121.0   81.0  28.73   
2                   0             0         0    245.0  127.5   80.0  25.34   
3                   0             1         0    225.0  150.0   95.0  28.58   
4                   0             0         0    285.0  130.0   84.0  23.10   
...               ...           ...       ...      ...    ...    ...    ...   
4233                0             1         0    313.0  179.0   92.0  25.97   
4234                0             0         0    207.0  126.5   80.0  19.71   
4235                0             0         0    248.0  131.0   72.0  22.00   
4236                0             0         0    210.0  126.5   87.0  19.16   
4237                0             0         0    269.0  133.5   83.0  21.47   

      heartRate  glucose  TenYearCHD  
0          80.0     77.0           0  
1          95.0     76.0           0  
2          75.0     70.0           0  
3          65.0    103.0           1  
4          85.0     85.0           0  
...         ...      ...         ...  
4233       66.0     86.0           1  
4234       65.0     68.0           0  
4235       84.0     86.0           0  
4236       86.0      NaN           0  
4237       80.0    107.0           0  

[4238 rows x 16 columns]

In [130]:
data.fillna(data.mean(), inplace=True)

data_array = data.values

In [131]:
data_array

array([[  1.        ,  39.        ,   4.        , ...,  80.        ,
         77.        ,   0.        ],
       [  0.        ,  46.        ,   2.        , ...,  95.        ,
         76.        ,   0.        ],
       [  1.        ,  48.        ,   1.        , ...,  75.        ,
         70.        ,   0.        ],
       ...,
       [  0.        ,  48.        ,   2.        , ...,  84.        ,
         86.        ,   0.        ],
       [  0.        ,  44.        ,   1.        , ...,  86.        ,
         81.96675325,   0.        ],
       [  0.        ,  52.        ,   2.        , ...,  80.        ,
        107.        ,   0.        ]])

### Task 1.2 - 1.3
Ajustar un modelo logístico polinomial en base al juego de datos cargado de forma matricial que relaciona las variables independientes que usted considere apropiadas (puede no utilizar todas las componentes de X), con la variable dependiente de salida (sufre o no sufre un paro cardíaco).   

Utilice la implementación vectorial del algoritmo de regresión logística (descenso del gradiente visto en clase).


In [132]:
# Componentes de X
X = data_array[:, 9:15]


In [133]:
X

array([[195.        , 106.        ,  70.        ,  26.97      ,
         80.        ,  77.        ],
       [250.        , 121.        ,  81.        ,  28.73      ,
         95.        ,  76.        ],
       [245.        , 127.5       ,  80.        ,  25.34      ,
         75.        ,  70.        ],
       ...,
       [248.        , 131.        ,  72.        ,  22.        ,
         84.        ,  86.        ],
       [210.        , 126.5       ,  87.        ,  19.16      ,
         86.        ,  81.96675325],
       [269.        , 133.5       ,  83.        ,  21.47      ,
         80.        , 107.        ]])

In [134]:
# Variable de salida
y = data_array[:, -1].astype(int)

In [135]:
y.reshape(y.shape[0], 1)

array([[0],
       [0],
       [0],
       ...,
       [0],
       [0],
       [0]])

In [136]:
X.shape

(4238, 6)

In [137]:
np.unique(X)

array([ 15.54,  15.96,  16.48, ..., 464.  , 600.  , 696.  ])

In [138]:
# Division en variables de entrenamiento y de testing

X_test, X_train = X[:len(X)//2], X[len(X)//2:]
print(X_test)
print(X_train)
Y_test, Y_train = y[:len(y)//2], y[len(y)//2:]
print(Y_test)
print(Y_train)

[[195.         106.          70.          26.97        80.
   77.        ]
 [250.         121.          81.          28.73        95.
   76.        ]
 [245.         127.5         80.          25.34        75.
   70.        ]
 ...
 [180.         113.          73.          17.65        70.
   73.        ]
 [177.         127.          79.          20.12        70.
   88.        ]
 [242.         115.          70.          32.07        83.
   81.96675325]]
[[258.         146.          98.5         26.05        60.
   68.        ]
 [218.         148.          80.          37.04        82.
   78.        ]
 [195.         137.          93.          26.39        88.
   75.        ]
 ...
 [248.         131.          72.          22.          84.
   86.        ]
 [210.         126.5         87.          19.16        86.
   81.96675325]
 [269.         133.5         83.          21.47        80.
  107.        ]]
[0 0 0 ... 0 0 0]
[0 0 0 ... 0 0 0]


In [139]:
def sigmoid(z):
    return 1 / (1 + np.exp(-z))

def logistic_regression(X, y, learning_rate=0.01, num_iterations=100):
    m, n = X.shape
    weights = np.zeros((n, 1))
    bias = 0
    for i in range(num_iterations):
        #Propagacion
        y_pred = sigmoid(np.dot(X, weights) + bias)

        # Costos
        cost = (-1/m) * np.sum(y * np.log(y_pred) + (1-y) * np.log(1-y_pred))

        
        # Calculo de gradiente
        dw = (1/m) * np.dot(X.T, (y_pred-y))
        db = (1/m) * np.sum(y_pred-y)

        

        weights = weights - learning_rate * dw
        bias = bias - learning_rate * db
        
        # Costos cada 100 iteraciones
        if i % 100 == 0:
            print("Cost despues de iteracion %i: %f" % (i, cost))

    return weights, bias

weights, bias = logistic_regression(X_train, Y_train, learning_rate=0.0001, num_iterations=100)

Cost despues de iteracion 0: 1468.778876


In [140]:
def predict(X, weights, bias):
    y_pred = sigmoid(np.dot(X, weights) + bias)

    y_pred[y_pred > 0.5] = 1
    y_pred[y_pred <= 0.5] = 0

    return y_pred

# Make predictions on the test set
y_pred = predict(X_test, weights, bias)

# Compute the accuracy
accuracy = (1 - np.mean(np.abs(y_pred - Y_test))) * 100

print("Accuracy:", accuracy, "%")

Accuracy: 74.32751297781972 %


### Task 1.4
Usando cross-validation determine el grado del polinomio que mejor describe la nube de puntos (encuentre el mejor balance entre apego a los datos de entrenamiento y generalización para datos previamente no observados).

In [148]:
def cross_validation(X, y, k=2, learning_rate=0.01, num_iters=100, degree_range = range(1,6)):
    m, n = X.shape
    fold_size = int(m/k)
    best_degree = None
    best_accuracy = 0
    
    for degree in degree_range:
        X_poly = np.ones((m, 1))
        for d in range(1, degree+1):
            X_poly = np.hstack((X_poly, np.power(X, d)))
            
        for i in range(k):
            X_val = X_poly[i*fold_size:(i+1)*fold_size, :]
            y_val = y[i*fold_size:(i+1)*fold_size]
            X_train = np.vstack((X_poly[:i*fold_size, :], X_poly[(i+1)*fold_size:, :]))
            y_train = np.concatenate((y[:i*fold_size], y[(i+1)*fold_size:]))
            
            weights, bias = logistic_regression(X_train, y_train, learning_rate, num_iters)
            
            y_pred = predict(X_val, weights, bias)
            
            accuracy = np.mean(y_pred == y_val)
            
            if accuracy > best_accuracy:
                best_degree = degree
                best_accuracy = accuracy
                
    return best_degree, best_accuracy  

In [149]:
best_degree, best_accuracy = cross_validation(X, y)

print('Best_degree: ', best_degree)
print('best_accuracy: ', best_accuracy)

Cost despues de iteracion 0: 1468.778876


C:\Users\Andre\AppData\Local\Temp\ipykernel_9136\95068840.py:2: RuntimeWarning: overflow encountered in exp
  return 1 / (1 + np.exp(-z))
C:\Users\Andre\AppData\Local\Temp\ipykernel_9136\95068840.py:13: RuntimeWarning: divide by zero encountered in log
  cost = (-1/m) * np.sum(y * np.log(y_pred) + (1-y) * np.log(1-y_pred))
C:\Users\Andre\AppData\Local\Temp\ipykernel_9136\95068840.py:13: RuntimeWarning: invalid value encountered in multiply
  cost = (-1/m) * np.sum(y * np.log(y_pred) + (1-y) * np.log(1-y_pred))


Cost despues de iteracion 0: 1468.778876
Cost despues de iteracion 0: 1468.778876
Cost despues de iteracion 0: 1468.778876
Cost despues de iteracion 0: 1468.778876
Cost despues de iteracion 0: 1468.778876
Cost despues de iteracion 0: 1468.778876
Cost despues de iteracion 0: 1468.778876
Cost despues de iteracion 0: 1468.778876
Cost despues de iteracion 0: 1468.778876
Best_degree:  1
best_accuracy:  0.7432751297781972


### Task 1.5 - Hallazgos

El uso del modelo de regresion logistica resulto ser moderadamente efectivo en cuanto a sus predicciones con los valores dados, como se puede observar, se obtuvo un 74% de aciertos en los datos de testeo posterior al entrenamiento del modelo. Por otro lado, la validacion cruzada nos ayuda a obtener la mejor variante de grado de polinomio, que nos ayude a representar de una mejor manera el modelo y reducir los posibles errores que se podrian llegar a presentar. Vale la pena mencionar que, la variable resultante es bastante dependiente del colesterol total que contiene asi como su nivel de glucosa y sus mediciones relacionadas al pulso del corazon, por lo que valdria la pena entrenar al modelo con datos similares para reducir las posibles incertidumbres acerca de las predicciones.

Recomendaciones para mejorar el modelo:
- Se podria utilizar una mayor cantidad de datos de entrenamiento al momento de entrenar el modelo, de esta manera se podra aumentar el porcentaje de precision del mismo.
- Se podrian utilizar otro tipo de variables (que posiblemente no esten listadas) que se correlacionen mas en cuanto si el paciente tiende mas a tener un infarto o no.
- Se podria mejorar la implementacion del modelo utilizando librerias externas que quizas tengan un mejor rendimiento que el actual.